### ISSUE TITLE
[maven-update] Automatically generated capsule code

### ISSUE THREAD

**Issue by dwtj** [dwtj/panini#3]
Does the automatically-generated capsule code get generated in the current commit (8f7af8f5)? It looks like it does not get generated, but I might be doing it wrong or looking in the wrong places. (The capsule `.class` files do not show up in `org.paninij/testing/target/classes/` and the capsule `.java` files do not show up in `org.paninij/testing/target/generated-sources/`.)

In any case, when the body of `org.paninij.lang.capsule.generation.TestHelloWorld` does not compile because it cannot find a `HelloWorldCapsule` class:

```
[ERROR] /Users/owl/Projects/panini/org.paninij/testing/src/test/java/org/paninij/lang/capsule/generation/TestHelloWorld.java:[8,9] cannot find symbol
[ERROR] symbol:   class HelloWorldCapsule
[ERROR] location: class org.paninij.lang.capsule.generation.TestHelloWorld
```

If support for capsule generation has not yet been added to the new maven build process, then how do we go about modifying the `pom.xml` files to support it?

**Comment by kristinclemens** [comment]
To my understanding it was build.xml that was actually doing the generation, right?

**Comment by dwtj** [comment]
That's right. Or maybe to be more precise, `build.xml` was describing the options that needed to be added to javac (or the eclipse compiler) in order to generate the code.

Here is the key section from `build.xml`:

```
    <javac target="1.8" srcdir="src" destdir="build/classes">
      <compilerarg value="-s"/>
      <compilerarg value="build/generated_src"/>
      <classpath path="lib/capsule-generator.jar"/>
    </javac>
```

The `compilerarg` tags are indicating where generated source files should go. We need something with equivalent functionality in the new maven-based build system.

**Comment by dwtj** [comment]
```
fir:~/Projects/panini% javac -help 
Usage: javac <options> <source files>
where possible options include:
  ...
  -s <directory>             Specify where to place generated source files
  ...
```

**Comment by kristinclemens** [comment]
Alright, I'm pretty sure I know how to make that happen. Maven has many [lifecycles](http://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#Lifecycle_Reference). Each lifecycle is composed of selection of build phases as you can see in that link. What we need to do is make sure that the `generate-sources` build phase knows how and where to build our sources. I'm going to figure that out right now.

**Comment by dwtj** [comment]
Ok. Do these stack overflow pages help?

[1](http://stackoverflow.com/questions/19633505/maven-generated-sources-not-compiled)
[2](http://stackoverflow.com/questions/4121010/maven-cant-add-files-in-generated-sources-for-compilation-phase)
[3](http://stackoverflow.com/questions/18721684/why-is-maven-skipping-over-my-custom-generate-sources-executions)

One of these suggests that that we might just need to add the generated-sources directory as a `<source>`.

**Comment by kristinclemens** [comment]
I edited my previous post to include the [glossary](http://maven.apache.org/archives/maven-1.x/reference/glossary.html).

**Comment by kristinclemens** [comment]
Yep, this is the bit we need right here:

    <build>
		<plugins>
			<plugin>
				<artifactId>maven-antrun-plugin</artifactId>
				<executions>
					<execution>
						<id>generate-sources</id>
						<phase>generate-sources</phase>
						<configuration>
							<tasks>
								<exec executable="${thrift.executable}">
									<arg value="--gen" />
									<arg value="java:beans" />
									<arg value="-o" />
									<arg value="src/main/java/com/... " />
									<arg value="src/main/thrift/... .thrift" />
								</exec>
							</tasks>
						</configuration>
						<goals>
							<goal>run</goal>
						</goals>
					</execution>
					<execution>
						<id>clean</id>
						<phase>clean</phase>
						<configuration>
							<tasks>
								<delete>
									<fileset dir="src/main/java/com/... " includes="... .java" />
								</delete>
							</tasks>
						</configuration>
						<goals>
							<goal>run</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
		</plugins>
    </build>

**Comment by dwtj** [comment]
That looks pretty different from our setup, doesn't it? We're not using Apache Thrift.

**Comment by kristinclemens** [comment]
Exactly, we'd still be using what WE'RE using, but the idea is that we can just use the Antrun maven plugin to do what we need to do.

**Comment by kristinclemens** [comment]
Using what you posted earlier,

    <javac target="1.8" srcdir="src" destdir="build/classes">
      <compilerarg value="-s"/>
      <compilerarg value="build/generated_src"/>
      <classpath path="lib/capsule-generator.jar"/>
    </javac>

I expect that, in our pom.xml for the paninij-testing project, we would have something like this:

    <exec executable="javac">
        <arg value="${paninij-lang-jar}" />
        ....
    </exec>

Yadda, yadda, as above in the current build.xml file. It's entirely possible we'll be able to just copy-paste that part of the code as-is from build.xml and just update the locations.

**Comment by dwtj** [comment]
I think that sources generated by the annotation processor are supposed to fit into the [compile phase](http://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html) rather than the generate-sources phase. That may sound counterintuitive, but I suspect that the generate-sources phase is meant to be used to hook into *external* source generation tools (i.e. *not* `javac`). I think that it is `compile` stage (when `javac` is run) that we want to configure. (Check out some of the options listed on the page that I linked to.)

**Comment by dwtj** [comment]
Of these options for `compiler:compile`, I think that we won't need to set `annotationProcessors` (because our `.jar` file will have put the processors in one of the "usual places"). Note also that we should not need to set the `generatedSourcesDirectory`, because it already has a sensible default. What do we need to set then?

**Comment by dwtj** [comment]
Hmm. Maybe we just need to pass the "-s" argument like [this](http://maven.apache.org/plugins/maven-compiler-plugin/examples/pass-compiler-arguments.html).

**Comment by dwtj** [comment]
I've been playing around with `testing/pom.xml`:

```
<project>
  ...
  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.2</version>
        <configuration>
          <source>1.8</source>
          <verbose>true</verbose>
          <fork>true</fork>
          <!--<compilerArgument>-s target/generated-sources</compilerArgument>-->
          <!--<proc>only</proc>-->
          <generatedSourcesDirectory>target/generated-sources</generatedSourcesDirectory>
        </configuration>
      </plugin>
    </plugins>
   
    ...
  </build>
```

After reviewing the default settings for both `compiler:compile` and `compiler:testCompile`, I think that we should not need to be changing the plugin's settings at all. I now think that the problem might be that the annotation processor *is not being found*.

**Comment by dwtj** [comment]
Yes, this is indeed the problem: the annotation processor service is not being found. CapsuleGenerator class is being packed into a `.jar` file correctly, but the `META-INF/service` information is not in the `.jar` file.

**Comment by dwtj** [comment]
When first tried to add the appropriate `org.paninij/lang/src/main/resources/META-INF/service/javax.annotation.processing.Processor` file, I could no longer compile `lang`.

Then I found [this](http://mail-archives.apache.org/mod_mbox/maven-users/200907.mbox/%3C24409779.post@talk.nabble.com%3E) interpretation of the error which I was seeing. By configuring the maven `compiler` plugin in `org.paninij/lang/pom.xml` to disable annotation processing *while building our annotation processor*, the compile phase of the `lang` module could be completed. With this change, the `lang` module can now produce a `.jar` file which includes the needed `META-INF/service` file.

Unfortunately, even after adding the service file to the `.jar` file, the `testing` module still does not build:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:testCompile (default-testCompile) on project testing: Compilation failure: Compilation failure:
[ERROR] /Users/owl/Projects/panini/org.paninij/testing/src/test/java/org/paninij/lang/capsule/generation/TestHelloWorld.java:[8,8] error: cannot find symbol
[ERROR] symbol:   class HelloWorldCapsule
[ERROR] location: class TestHelloWorld
[ERROR] /Users/owl/Projects/panini/org.paninij/testing/src/test/java/org/paninij/lang/capsule/generation/TestHelloWorld.java:[8,36] error: cannot find symbol
[ERROR] -> [Help 1]
```

It still seems that the `testing` module doesn't run it the `CapsuleGenerator` processor during its compile phase.

:(

**Comment by dwtj** [comment]
I just tested to make sure that the `CapsuleGenerator` class is being found when trying to compile the `testing` module:

```
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.2</version>
        <configuration>
          <source>1.8</source>
          <verbose>true</verbose>
          <fork>true</fork>
          <!--<proc>only</proc>-->
          <generatedSourcesDirectory>target/generated-sources</generatedSourcesDirectory>
          <annotationProcessors>
            <annotationProcessor>org.paninij.lang.capsule.generation.CapsuleGenerator</annotationProcessor>
            <!--<annotationProcessor>org.paninij.lang.capsule.generation.CapsuleGeneratorDNE</annotationProcessor>-->
          </annotationProcessors>
        </configuration>
      </plugin>
    </plugins>
```

When the second annotation processor is included, the build of `lang` fails because that class cannot be found. This kind of error does not occur when only the first annotation processor is included.

**Comment by dwtj** [comment]
Two posts previously, I described disabling annotation processing while compiling the `lang` module in order to make it work. I believe that this was misguided, because annotation processing is *required* to create annotations. (See the current definitions of `Capsule` and `CapsuleGenerator`.)

To reiterate: the current maven build process does not generate a `.jar` file which includes the needed service file. When I attempt to add the this service file just the same way that we did in the previous `capsule-generator` maven build system, the `lang` module no longer compiles.

**Comment by dwtj** [comment]
I fixed some obvious bugs in commit 89698794a9bf0ee06be647714b7b793724551dbf. I have been able to get this working, though not consistently. I have, on occasion, been able to build valid `.jar` files which I can use with the (`ant`-based) `capsule-generator-tests` repository. The build process for `lang` seems to be very temperamental, however, and I have no idea why. When the `lang` module fails to build, it is still because of the "Bad service configuration file" error:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:compile (default-compile) on project lang: Compilation failure
[ERROR] Bad service configuration file, or exception thrown while constructing Processor object: javax.annotation.processing.Processor: Provider org.paninij.lang.capsule.generation.CapsuleGenerator not found
```

This error even came up when I tried to exclude this service resource from the build in the `lang/pom.xml`.

**Comment by dwtj** [comment]
It looks like [this](https://cdivilly.wordpress.com/2010/03/16/maven-and-jsr-269-annotation-processors/) is describing the problem that I am having:

> In a Maven project the logical placement of this file is the src/main/resources folder. With the default maven lifecycle the contents of this folder are copied to the output folder (which forms part of the classpath searched by the compiler for annotation processors) before compilation takes place. This causes the compiler to expect a compiled instance of annotation processors enumerated in the file to be available at compilation time. If the compiler fails to find the processors it will raise an error, causing the build to abort. Of course in the project where the annotation processors are defined this causes a chicken and egg problem. The processors need to be first compiled by the compiler but the compiler expects them to already exist.
>
> The workaround is to modify the maven build lifecycle to compile the project in two passes. The first pass compiles just the annotation processors (with annotation processing disabled), the second compiles the rest of the project (with annotation processing enabled). The first pass makes the compiled annotation processors available for the second pass.

This site also proposes a solution to the problem, but I think the solution described [here in the comments](https://community.oracle.com/thread/1184021) to be simpler: make the annotation processors be in their own module and compile them in a forked `javac` without processors turned on.

**Comment by dwtj** [comment]
Fixed as of commit d246719712479832d1356303d63675d66c5fd17b (done on a new branch [maven-modules] rather than the original [maven-update]).
